csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll

```diff
@@ -0,0 +1,998 @@
+
+  override Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    tag = InitializerVariableAddressTag() and
+    result = getInitialization().getFirstInstruction() and
+    kind instanceof GotoEdge
+  }
+
+  override Instruction getChildSuccessor(TranslatedElement child) {
+    child = getInitialization() and
+    result = getEnclosingFunction().getReturnSuccessorInstruction()
+  }
+
+  override IRVariable getInstructionVariable(InstructionTag tag) {
+    tag = InitializerVariableAddressTag() and
+    result = getEnclosingFunction().getReturnVariable()
+  }
+
+  override Instruction getTargetAddress() {
+    result = getInstruction(InitializerVariableAddressTag())
+  }
+
+  override Type getTargetType() {
+    result = getEnclosingFunction().getReturnVariable().getType()
+  }
+
+  TranslatedInitialization getInitialization() {
+    result = getTranslatedInitialization(
+      stmt.getExpr())
+  }
+}
+
+class TranslatedReturnVoidStmt extends TranslatedReturnStmt {
+  TranslatedReturnVoidStmt() {
+    not exists(stmt.getExpr())
+  }
+
+  override TranslatedElement getChild(int id) {
+    none()
+  }
+
+  override Instruction getFirstInstruction() {
+    result = getInstruction(OnlyInstructionTag())
+  }
+
+  override predicate hasInstruction(Opcode opcode, InstructionTag tag,
+    Type resultType, boolean isLValue) {
+    tag = OnlyInstructionTag() and
+    opcode instanceof Opcode::NoOp and
+    resultType instanceof VoidType and
+    isLValue = false
+  }
+
+  override Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    tag = OnlyInstructionTag() and
+    result = getEnclosingFunction().getReturnSuccessorInstruction() and
+    kind instanceof GotoEdge
+  }
+
+  override Instruction getChildSuccessor(TranslatedElement child) {
+    none()
+  }
+}
+
+/**
+ * The IR translation of a C++ `try` statement.
+ */
+class TranslatedTryStmt extends TranslatedStmt {
+  override TryStmt stmt;
+
+  override TranslatedElement getChild(int id) {
+    id = 0 and result = getBody() or
+    id = 1 and result = getFinally() or
+    result = getCatchClause(id - 2)
+  }
+
+  override predicate hasInstruction(Opcode opcode, InstructionTag tag,
+    Type resultType, boolean isLValue) {
+    none()   
+  }
+
+  override Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    none()  
+  }
+
+  override Instruction getFirstInstruction() {
+    result = getBody().getFirstInstruction()
+  }
+
+  override Instruction getChildSuccessor(TranslatedElement child) {
+    child = getCatchClause(_) and result = getFinally().getFirstInstruction() or
+    child = getBody() and result = getFinally().getFirstInstruction() or
+    child = getFinally() and result = getParent().getChildSuccessor(this)
+  }
+
+  final Instruction getNextHandler(TranslatedClause clause) {
+    exists(int index |
+      clause = getCatchClause(index) and
+      result = getCatchClause(index + 1).getFirstInstruction()
```

Comment:
We also have `when` clauses to consider.

Reply:
Indeed, I think the `if` - `else` desugaring would be a better fit when `when` clauses will be implemented.

Reply:
Sure, let's defer it. Just mentioning in passing.